cranelift/wasm/src/code_translator.rs

```diff
@@ -2542,6 +2574,91 @@ pub fn optionally_bitcast_vector(
     }
 }
 
+#[inline(always)]
+fn is_non_canonical_v128(ty: ir::Type) -> bool {
+    match ty {
+        B8X16 | B16X8 | B32X4 | B64X2 | I64X2 | I32X4 | I16X8 | F32X4 | F64X2 => true,
+        _ => false,
+    }
+}
+
+/// Cast to I8X16, any vector values in `values` that are of "non-canonical" type (meaning, not
+/// I8X16) and park them in the supplied smallvec `canonicalised`.  But before doing that, make
+/// a pre-scan to check if any of this is necessary, and return `false` if not.
+fn canonicalise_v128_values(
+    canonicalised: &mut SmallVec<[ir::Value; 16]>,
+    builder: &mut FunctionBuilder,
+    values: &[ir::Value],
+) -> bool {
+    debug_assert!(canonicalised.is_empty());
+    // First figure out if any of the parameters need to be cast.  Mostly they don't need to be.
+    let mut any_non_canonical = false;
+    for v in values {
+        if is_non_canonical_v128(builder.func.dfg.value_type(*v)) {
+            any_non_canonical = true;
+            break;
+        }
+    }
+    // Hopefully we take this exit most of the time, hence doing no heap allocation.
+    if !any_non_canonical {
+        return false;
+    }
+    // Otherwise we'll have to cast, and push the resulting `Value`s into `canonicalised`.
+    for v in values {
+        canonicalised.push(if is_non_canonical_v128(builder.func.dfg.value_type(*v)) {
+            builder.ins().raw_bitcast(I8X16, *v)
+        } else {
+            *v
+        });
+    }
+    // Let the caller know we had to copy.
+    true
+}
+
+/// Generate a `jump` instruction, but first cast all 128-bit vector values to I8X16 if they
+/// don't have that type.  This is done in somewhat roundabout way so as to ensure that we
+/// almost never have to do any heap allocation.
+fn canonicalise_then_jump(
+    builder: &mut FunctionBuilder,
+    destination: ir::Block,
+    params: &[ir::Value],
+) -> ir::Inst {
+    let mut canonicalised = SmallVec::<[ir::Value; 16]>::new();
+    if canonicalise_v128_values(&mut canonicalised, builder, params) {
+        builder.ins().jump(destination, canonicalised.as_slice())
+    } else {
+        builder.ins().jump(destination, params)
+    }
+}
+
+/// The same but for a `brz` or `brnz` instruction.
+fn canonicalise_then_br_z_or_nz(
+    is_z: bool,
```

Comment:
let's split that into two functions: `canonicalise_then_br_z` and `canonicalise_then_br_nz`